pkg/sources/filesystem/filesystem.go

```diff
@@ -181,23 +182,31 @@ func (s *Source) scanFile(ctx context.Context, path string, chunksChan chan *sou
 		return err
 	}
 	reReader.Stop()
-	data, err := io.ReadAll(reReader)
-	if err != nil {
-		return fmt.Errorf("unable to read file: %w", err)
-	}
-	chunksChan <- &sources.Chunk{
-		SourceType: s.Type(),
-		SourceName: s.name,
-		SourceID:   s.SourceID(),
-		Data:       data,
-		SourceMetadata: &source_metadatapb.MetaData{
-			Data: &source_metadatapb.MetaData_Filesystem{
-				Filesystem: &source_metadatapb.Filesystem{
-					File: sanitizer.UTF8(path),
+
+	for {
+		chunkBytes := make([]byte, BufferSize)
+		reader := bufio.NewReaderSize(reReader, BufferSize)
+		n, err := reader.Read(chunkBytes)
+		if err != nil && !errors.Is(err, io.EOF) {
```

Comment:
I believe if reader.Read doesn't read anything, it returns an EOF. So how does this loop terminate? Or am I wrong?

Reply:
You're correct. Blew through this too late on a Friday and counted on the test too much. Going to get a better filesystem test added.

Reply:
I read through the [Reader](https://pkg.go.dev/io#Reader) interface docs, which recommends:

> Callers should always process the n > 0 bytes returned before considering the error err. Doing so correctly handles I/O errors that happen after reading some bytes and also both of the allowed EOF behaviors.

So that would suggest checking `err != nil` *after* our `n > 0` expression.

I think I would feel safer with something like that and:
```go
if err != nil || n == 0 {
    break
}
```

Which begs the question, should we return the `err` if it's not an EOF error?

Reply:
I'm ok with switching, but this should work too. The earlier check if the error is not EOF and the one at the end checking if it is EOF is to cover the two behaviors (hits end of reader w/ EOF or hits end of reader and returns EOF on next run iteration).